manta-pay/src/crypto/poseidon/mod.rs

```diff
@@ -0,0 +1,447 @@
+            self.full_round(round, state, compiler);
+        }
+        for round in Self::HALF_FULL_ROUNDS..(Self::HALF_FULL_ROUNDS + S::PARTIAL_ROUNDS) {
+            self.partial_round(round, state, compiler);
+        }
+        for round in
+            (Self::HALF_FULL_ROUNDS + S::PARTIAL_ROUNDS)..(S::FULL_ROUNDS + S::PARTIAL_ROUNDS)
+        {
+            self.full_round(round, state, compiler);
+        }
+    }
+
+    /// Computes the first round borrowing the `input` and `domain_tag` returning the [`State`]
+    /// after the first round. This method does not check that `N + 1 = S::WIDTH`.
+    #[inline]
+    fn first_round_with_domain_tag_unchecked<const N: usize>(
+        &self,
+        domain_tag: &S::Field,
+        input: [&S::Field; N],
+        compiler: &mut COM,
+    ) -> State<S, COM> {
+        let mut state = Vec::with_capacity(S::WIDTH);
+        for (i, point) in iter::once(domain_tag).chain(input).enumerate() {
+            let mut elem = S::add_const(point, &self.additive_round_keys[i], compiler);
+            S::apply_sbox(&mut elem, compiler);
+            state.push(elem);
+        }
+        let mut state = State(state.into_boxed_slice());
+        self.mds_matrix_multiply(&mut state, compiler);
+        state
+    }
+}
+
+impl<S, COM> Decode for Permutation<S, COM>
+where
+    S: Specification<COM>,
+    S::ParameterField: Decode,
+{
+    type Error = <S::ParameterField as Decode>::Error;
+
+    #[inline]
+    fn decode<R>(mut reader: R) -> Result<Self, DecodeError<R::Error, Self::Error>>
+    where
+        R: Read,
+    {
+        Ok(Self::new_unchecked(
+            (0..Self::ADDITIVE_ROUND_KEYS_COUNT)
+                .map(|_| Decode::decode(&mut reader))
+                .collect::<Result<_, _>>()?,
+            (0..Self::MDS_MATRIX_SIZE)
+                .map(|_| Decode::decode(&mut reader))
+                .collect::<Result<_, _>>()?,
+        ))
+    }
+}
+
+impl<S, COM> Encode for Permutation<S, COM>
+where
+    S: Specification<COM>,
+    S::ParameterField: Encode,
+{
+    #[inline]
+    fn encode<W>(&self, mut writer: W) -> Result<(), W::Error>
+    where
+        W: Write,
+    {
+        for key in self.additive_round_keys.iter() {
+            key.encode(&mut writer)?;
+        }
+        for entry in self.mds_matrix.iter() {
+            entry.encode(&mut writer)?;
+        }
+        Ok(())
+    }
+}
+
+impl<S, COM> PseudorandomPermutation<COM> for Permutation<S, COM>
+where
+    S: Specification<COM>,
+{
+    type Domain = State<S, COM>;
+
+    #[inline]
+    fn permute(&self, state: &mut Self::Domain, compiler: &mut COM) {
+        self.full_round(0, state, compiler);
+        self.permute_without_first_round(state, compiler);
+    }
+}
+
+impl<D, S, COM> Sample<D> for Permutation<S, COM>
+where
+    D: Clone,
+    S: Specification<COM>,
+    S::ParameterField: Field + FieldGeneration + PartialEq + Sample<D>,
+{
+    /// Samples random Poseidon parameters.
```

Comment:
Shall we remove this doc? This is trait implementation

Reply:
Shall we remove this doc? This is trait implementation

Reply:
Yes. Should be removed.